### PR TITLE
Social auth

### PR DESCRIPTION
**Pull request by lady-ann**

### CODE REVIEW

File: src/main/kotlin/dev/tmsoft/lib/socialauth/SocialAuth.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+fun transformSocial(
+    pipeline: ApplicationCallPipeline,
+    transformer: SocialPrincipalTransformer
+) {
+    val phase = PipelinePhase("SocialAuth")
+    pipeline.insertPhaseAfter(Authentication.ChallengePhase, phase)
+    pipeline.intercept(phase) {
+        val oauthPrincipal = call.authentication.principal<OAuthAccessTokenResponse>()
+        if (oauthPrincipal != null) {
+            val principal = transformer.transform(oauthPrincipal)
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/socialauth/SocialAuth.kt:65]: save principal in the call `call.authentication.principal = principal`

File: src/main/kotlin/dev/tmsoft/lib/socialauth/TwitterAPI.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import javax.crypto.spec.SecretKeySpec
+import io.ktor.client.request.get
+import io.ktor.client.request.headers
+import io.ktor.http.HttpHeaders
+import kotlinx.serialization.json.Json
+
+private const val USER_URL = "https://api.twitter.com/1.1/account/verify_credentials.json"
+
+class TwitterAPI(private val clientKey: String, private val clientSecret: String) {
+    private val client = HttpClient(CIO) {
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/socialauth/TwitterAPI.kt:24]: make a single client for all networks

File: src/main/kotlin/dev/tmsoft/lib/socialauth/TwitterAPI.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
+package dev.tmsoft.lib.socialauth
+
+import io.ktor.client.HttpClient
+import io.ktor.client.engine.cio.CIO
```
Comment (shustrik) [inline src/main/kotlin/dev/tmsoft/lib/socialauth/TwitterAPI.kt:4]: add CIO dependency into build.gradle.kts

**Review by shustrik — CHANGES_REQUESTED** [review]

File: src/main/kotlin/dev/tmsoft/lib/socialauth/SocialAuth.kt (hunk 1 above)
Comment (lady-ann) [reply src/main/kotlin/dev/tmsoft/lib/socialauth/SocialAuth.kt:65]: Principal could be saved only once per call

**Review by shustrik — APPROVED** [review]

**Pull request merged**